feat(gic_v2): add GIC Hypervisor Interface support

luodeb opened the pull request:
- Implement GicHypervisorInterface struct and related methods
- Add new GIC registers for hypervisor interface
- Update lib.rs to include GicHypervisorInterface in public exports
- Introduce GIC_LIST_REGS_NUM constant for list register count

luodeb committed:
feat(gic_v2): add GIC Hypervisor Interface support

- Implement GicHypervisorInterface struct and related methods
- Add new GIC registers for hypervisor interface
- Update lib.rs to include GicHypervisorInterface in public exports
- Introduce GIC_LIST_REGS_NUM constant for list register count

hky1999 commented on src/gic_v2.rs:
```diff
@@ -87,6 +89,36 @@ register_structs! {
+register_structs! {
+    /// GIC Hypervisor Interface registers.
+    #[allow(non_snake_case)]
+    GicHypervisorInterfaceRegs {
```
you can introduce `gich` feature

luodeb replied:
ok

hky1999 commented on src/gic_v2.rs:
```diff
@@ -465,3 +513,126 @@ impl GicCpuInterface {
+    ///
+    /// The number of List Registers supported by the GIC, derived from the VTR register.
+    pub fn get_vtr(&self) -> u32 {
+        let vtr: u32 = self.regs().VTR.get();
```
refactor `VTR` like https://github.com/arceos-hypervisor/arm_gicv2/blob/main/src/regs/gicd_sgir.rs

for better readability

luodeb replied:
ok

hky1999 commented on src/lib.rs:
```diff
@@ -7,7 +7,7 @@ use core::ops::Range;
 mod gic_v2;
 
-pub use gic_v2::{GicCpuInterface, GicDistributor};
+pub use gic_v2::{GicCpuInterface, GicDistributor, GicHypervisorInterface};
```
#[cfg(feature = "hv")]

pub use gic_v2::GicHypervisorInterface;

luodeb replied:
ok

luodeb committed:
feat(regs): add support for GIC Hypervisor Control Registers

- Add new register definitions for GIC Hypervisor Control Registers
- Update GicHypervisorInterfaceRegs to use new register types
- Add new modules for GIC Hypervisor Control Registers in regs directory
- Update mod.rs to include new register modules

luodeb committed:
feat(hv): add hypervisor support for GICv2

- Add GicHypervisorInterface struct and related functionality
- Implement methods for accessing hypervisor-specific GIC registers
- Add support for virtual interrupt management and maintenance interrupts
- Update Cargo.toml to include 'hv' feature
- Refactor gic_v2 module to separate hypervisor-related code

luodeb committed:
format code.

hky1999 commented on src/gich.rs:
```diff
@@ -0,0 +1,181 @@
+/// GIC Hypervisor Interface
```
//!

luodeb replied:
ok

hky1999 commented on src/regs/gich/gich_apr.rs:
```diff
@@ -0,0 +1,163 @@
+    pub GICH_APR [
+        /// [31:0] P<x>
+        /// Active priorities. Each bit indicates whether there is an interrupt active at the priority corresponding to that bit.
+        INTERRUPT31 OFFSET(31) NUMBITS(1) [
```
do we need to seperate each "bit" for this u32 bit map?

or simply use `ACTIVE_PRIORITY_BITS OFFSET(0) NUMBITS(32)` ?

luodeb replied:
ok

hky1999 commented on src/regs/gich/gich_eisr.rs:
```diff
@@ -0,0 +1,97 @@
+        /// [15:0] Status<n>
+        /// EOI maintenance interrupt status for List register <n>.
+        // Status [
+            Status15 OFFSET(15) NUMBITS(1) [
```
same question as `GICH_APR`, do we need to seperate each bits?

it depends on our operation on this fields. in my opinion this will make the function more complex.

luodeb replied:
OK

hky1999 commented on src/regs/gich/gich_elrsr.rs:
```diff
@@ -0,0 +1,97 @@
+        /// [15:0] Status<n>
+        /// Status bit for List register <n>.
+        // Status [
+            Status15 OFFSET(15) NUMBITS(1) [
```
same question as `GICH_APR`

luodeb replied:
OK

hky1999 commented on src/gich.rs:
```diff
@@ -0,0 +1,181 @@
+    /// # Returns
+    ///
+    /// The current value of the APR register.
+    pub fn apr(&self) -> u32 {
```
how to use this api when you want to get if a target preemption level (0-31) is active or not?

luodeb replied:
`(GICH.apr() & (1 << bit_offset)) != 0` is a good usage.

luodeb committed:
refactor(gich): simplify register bitfields and improve documentation

- Simplified bitfield definitions for GICH_APR, GICH_EISR, and GICH_ELRSR registers
- Replaced individual interrupt status fields with consolidated bit ranges
- Updated module documentation to use more appropriate formatting
- Improved readability and maintainability of register definitions

luodeb committed:
code format.
